src/System.Data.Common/src/System/Data/Common/DbProviderFactories.cs

```diff
@@ -0,0 +1,187 @@
+                    }
+                    throw ADP.Argument(SR.Format(SR.ADP_DbProviderFactories_FactoryNotLoadable, assemblyQualifiedName));
+                }
+            }
+            throw ADP.Argument(SR.ADP_DbProviderFactories_NoAssemblyQualifiedName);
+        }
+
+        public static DataTable GetFactoryClasses()
+        {
+            try
+            {
+                _providerTableLock.EnterReadLock();
+                return _providerTable == null ? GetInitialProviderTable() : _providerTable.Copy();
+            }
+            finally
+            {
+                _providerTableLock.ExitReadLock();
+            }
+        }
+
+        public static DbProviderFactory GetFactory(DbConnection connection)
+        {
+            ADP.CheckArgumentNull(connection, nameof(connection));
+            return connection.ProviderFactory;
+        }
+
+        public static void ConfigureFactory(Type providerFactoryClass, string providerInvariantName)
+        {
+            ConfigureFactory(providerFactoryClass, providerInvariantName, string.Empty, string.Empty);
+        }
+        
+        public static void ConfigureFactory(Type providerFactoryClass, string providerInvariantName, string name, string description)
+        {
+            ADP.CheckArgumentNull(providerFactoryClass, nameof(providerFactoryClass));
+            ADP.CheckArgumentLength(providerInvariantName, nameof(providerInvariantName));
+
+            if (!providerFactoryClass.IsSubclassOf(typeof(DbProviderFactory)))
+            {
+                throw ADP.Argument(SR.Format(SR.ADP_DbProviderFactories_NotAFactoryType, providerFactoryClass.FullName));
+            }
+            RegisterFactoryInTable(providerFactoryClass, providerInvariantName, name, description);
+        }
+        
+        public static void ConfigureFactory(DbConnection connection, string providerInvariantName)
+        {
+            ConfigureFactory(connection, providerInvariantName, string.Empty, string.Empty);
+        }
+        
+        public static void ConfigureFactory(DbConnection connection, string providerInvariantName, string name, string description)
+        {
+            ADP.CheckArgumentNull(connection, nameof(connection));
+            ADP.CheckArgumentLength(providerInvariantName, nameof(providerInvariantName));
+
+            DbProviderFactory factoryInstance = GetFactory(connection);
+            if (factoryInstance == null)
+            {
+                throw ADP.Argument(SR.ADP_DbProviderFactories_NoFactorySuppliedThroughDbConnection);
+            }
+            RegisterFactoryInTable(factoryInstance.GetType(), providerInvariantName, name, description);
+        }
+        
+        private static void RegisterFactoryInTable(Type factoryType, string providerInvariantName, string name, string description)
+        {
+            ADP.CheckArgumentNull(factoryType, nameof(factoryType));
```

Comment:
@divega At the moment it appears it's not used indeed. It's fine by me to remove it, or replace it with a Debug.Assert, whatever makes you all happy :)

> And while we are on it, can we switch to var everything that can be switched to var in this code?

I'd love to, I followed the coding guidelines which explicitly stated 'var' was out of the question and should be avoided, so I used the types instead :)